In [7]:
import json
import pandas as pd
import ast
import math

In [8]:
all_nodes_path = "C:/Users/islam/Downloads/all_contract_text.json"
highlighted_nodes_path = "C:/Users/islam/Downloads/contract_saved.json"
with open(all_nodes_path, encoding='UTF-8') as f:
    all_nodes_data = json.load(f)
with open(highlighted_nodes_path, encoding='UTF-8') as f:
    highlighted_data = json.load(f)

In [9]:
all_nodes_xpaths = ast.literal_eval(all_nodes_data['xpaths'])
all_nodes_segmented_text = ast.literal_eval(all_nodes_data['segmentedTexts'])
df = pd.DataFrame()
df['xpaths'] = all_nodes_xpaths
df['text'] = all_nodes_segmented_text
df['all_nodes_ordering'] = df.index.copy()

In [10]:
highlighted_data['c']

'[["84.3125px","543.5px","816px","175px"],["313.625px","880.391px","142.219px","15px"],["1730.12px","943.344px","16.2969px","15px"],["978.125px","945.203px","12.5938px","15px"],["1853.12px","917.906px","67.1719px","15px"],["2899.12px","943.719px","15.5469px","15px"],["2456.12px","880.406px","142.172px","15px"],["2424.12px","918.75px","65.4844px","15px"],["2488.12px","643.609px","106.656px","15px"],["2488.12px","616.938px","23.3281px","15px"],["3473.12px","943.719px","15.5469px","15px"],["4549.12px","943.719px","15.5469px","15px"],["4015.12px","943.719px","15.5469px","15px"],["5075.12px","943.719px","15.5469px","15px"],["8316.12px","909.094px","84.7969px","15px"],["8284.12px","918.75px","65.4844px","15px"],["8348.12px","643.609px","66.1562px","15px"],["8348.12px","616.938px","26.6719px","15px"],["8447.12px","940.391px","22.2188px","15px"],["8562.12px","635.078px","111.312px","15px"],["8562.12px","616.938px","14.7969px","15px"],["8490.12px","635.828px","109.219px","15px"],["8490.12px","6

In [11]:
highlight_xpaths = ast.literal_eval(highlighted_data['xpaths'])
highlight_segmented_text = ast.literal_eval(highlighted_data['segmentedTexts'])
highlight_text = ast.literal_eval(highlighted_data['texts'])
highlight_labels = ast.literal_eval(highlighted_data['labels'])
highlight_coordinates = ast.literal_eval(highlighted_data['c'])
highlighted_df = pd.DataFrame()
highlighted_df['highlighted_xpaths'] = highlight_xpaths
highlighted_df['highlighted_segmented_text'] = highlight_segmented_text
highlighted_df['highlighted_labels'] = highlight_labels
highlighted_df['highlighted_coordinates'] = highlight_coordinates
highlighted_df['segment_number_from_idx'] = highlighted_df.index.copy()
highlighted_df['num_entries_1'] = highlighted_df['highlighted_xpaths'].apply(len)
highlighted_df['num_entries_2'] = highlighted_df['highlighted_segmented_text'].apply(len)

In [12]:
assert highlighted_df['num_entries_1'].equals(highlighted_df['num_entries_2']), 'Mismatch in segmentation and groupings'
print("There is no mismatch in segmentations and groupings, we can proceed")

There is no mismatch in segmentations and groupings, we can proceed


In [60]:
exploded_highlight_df = highlighted_df[
    ['highlighted_xpaths','highlighted_segmented_text','highlighted_labels',
     'segment_number_from_idx','highlighted_coordinates','num_entries_1']
    ].explode(column=['highlighted_xpaths','highlighted_segmented_text']).reset_index(drop=True)

exploded_highlight_df['exploded_highlight_node_order'] = exploded_highlight_df.index.copy()

In [61]:
exploded_highlight_df

highlighted_xpaths  \
0   /html/body/document/type/sequence/filename/des...   
1   /html/body/document/type/sequence/filename/des...   
2   /html/body/document/type/sequence/filename/des...   
3   /html/body/document/type/sequence/filename/des...   
4   /html/body/document/type/sequence/filename/des...   
5   /html/body/document/type/sequence/filename/des...   
6   /html/body/document/type/sequence/filename/des...   
7   /html/body/document/type/sequence/filename/des...   
8   /html/body/document/type/sequence/filename/des...   
9   /html/body/document/type/sequence/filename/des...   
10  /html/body/document/type/sequence/filename/des...   
11  /html/body/document/type/sequence/filename/des...   
12  /html/body/document/type/sequence/filename/des...   
13  /html/body/document/type/sequence/filename/des...   
14  /html/body/document/type/sequence/filename/des...   
15  /html/body/document/type/sequence/filename/des...   
16  /html/body/document/type/sequence/filename/des...   
17  /html/body/document/type/sequence/filename/des...   
18  /html/body/document/type/sequence/filename/des...   
19  /html/body/document/type/sequence/filename/des...   
20  /html/body/document/type/sequence/filename/des...   
21  /html/body/document/type/sequence/filename/des...   
22  /html/body/document/type/sequence/filename/des...   
23  /html/body/document/type/sequence/filename/des...   
24  /html/body/document/type/sequence/filename/des...   
25  /html/body/document/type/sequence/filename/des...   
26  /html/body/document/type/sequence/filename/des...   
27  /html/body/document/type/sequence/filename/des...   
28  /html/body/document/type/sequence/filename/des...   
29  /html/body/document/type/sequence/filename/des...   
30  /html/body/document/type/sequence/filename/des...   
31  /html/body/document/type/sequence/filename/des...   
32  /html/body/document/type/sequence/filename/des...   
33  /html/body/document/type/sequence/filename/des...   
34  /html/body/document/type/sequence/filename/des...   
35  /html/body/document/type/sequence/filename/des...   
36  /html/body/document/type/sequence/filename/des...   
37  /html/body/document/type/sequence/filename/des...   

             highlighted_segmented_text highlighted_labels  \
0          AGREEMENT AND PLAN OF MERGER                  t   
1            dated as of March 22, 2021                  t   
2                        by and between                  t   
3             BANC OF\nCALIFORNIA, INC.                  t   
4                                   and                  t   
5            PACIFIC MERCANTILE BANCORP                  t   
6                     TABLE OF CONTENTS                 st   
7                                  -ii-                  n   
8                                   -i-                  n   
9                              RECITALS                 st   
10                                  -2-                  n   
11                  CERTAIN DEFINITIONS                 st   
12                            ARTICLE 1                 sn   
13                  Certain Definitions                sst   
14                                    .                sst   
15                                 1.01                ssn   
16                                  -3-                  n   
17                                  -5-                  n   
18                                  -4-                  n   
19                                  -6-                  n   
20                           THE MERGER                 st   
21                            ARTICLE 2                 sn   
22                          The\nMerger                 st   
23                                    .                 st   
24                                 2.01                 sn   
25                                 -12-                  n   
26                 Effect of the Merger               ssst   
27                                    .               ssst   
28           

In [62]:
exploded_highlight_df['top'] = exploded_highlight_df['highlighted_coordinates'].apply(lambda x: float(x[0][:-2]))
exploded_highlight_df['left'] = exploded_highlight_df['highlighted_coordinates'].apply(lambda x: float(x[1][:-2]))
exploded_highlight_df['width'] = exploded_highlight_df['highlighted_coordinates'].apply(lambda x: float(x[2][:-2]))
exploded_highlight_df['height'] = exploded_highlight_df['highlighted_coordinates'].apply(lambda x: float(x[3][:-2]))

In [63]:
exploded_highlight_df

highlighted_xpaths  \
0   /html/body/document/type/sequence/filename/des...   
1   /html/body/document/type/sequence/filename/des...   
2   /html/body/document/type/sequence/filename/des...   
3   /html/body/document/type/sequence/filename/des...   
4   /html/body/document/type/sequence/filename/des...   
5   /html/body/document/type/sequence/filename/des...   
6   /html/body/document/type/sequence/filename/des...   
7   /html/body/document/type/sequence/filename/des...   
8   /html/body/document/type/sequence/filename/des...   
9   /html/body/document/type/sequence/filename/des...   
10  /html/body/document/type/sequence/filename/des...   
11  /html/body/document/type/sequence/filename/des...   
12  /html/body/document/type/sequence/filename/des...   
13  /html/body/document/type/sequence/filename/des...   
14  /html/body/document/type/sequence/filename/des...   
15  /html/body/document/type/sequence/filename/des...   
16  /html/body/document/type/sequence/filename/des...   
17  /html/body/document/type/sequence/filename/des...   
18  /html/body/document/type/sequence/filename/des...   
19  /html/body/document/type/sequence/filename/des...   
20  /html/body/document/type/sequence/filename/des...   
21  /html/body/document/type/sequence/filename/des...   
22  /html/body/document/type/sequence/filename/des...   
23  /html/body/document/type/sequence/filename/des...   
24  /html/body/document/type/sequence/filename/des...   
25  /html/body/document/type/sequence/filename/des...   
26  /html/body/document/type/sequence/filename/des...   
27  /html/body/document/type/sequence/filename/des...   
28  /html/body/document/type/sequence/filename/des...   
29  /html/body/document/type/sequence/filename/des...   
30  /html/body/document/type/sequence/filename/des...   
31  /html/body/document/type/sequence/filename/des...   
32  /html/body/document/type/sequence/filename/des...   
33  /html/body/document/type/sequence/filename/des...   
34  /html/body/document/type/sequence/filename/des...   
35  /html/body/document/type/sequence/filename/des...   
36  /html/body/document/type/sequence/filename/des...   
37  /html/body/document/type/sequence/filename/des...   

             highlighted_segmented_text highlighted_labels  \
0          AGREEMENT AND PLAN OF MERGER                  t   
1            dated as of March 22, 2021                  t   
2                        by and between                  t   
3             BANC OF\nCALIFORNIA, INC.                  t   
4                                   and                  t   
5            PACIFIC MERCANTILE BANCORP                  t   
6                     TABLE OF CONTENTS                 st   
7                                  -ii-                  n   
8                                   -i-                  n   
9                              RECITALS                 st   
10                                  -2-                  n   
11                  CERTAIN DEFINITIONS                 st   
12                            ARTICLE 1                 sn   
13                  Certain Definitions                sst   
14                                    .                sst   
15                                 1.01                ssn   
16                                  -3-                  n   
17                                  -5-                  n   
18                                  -4-                  n   
19                                  -6-                  n   
20                           THE MERGER                 st   
21                            ARTICLE 2                 sn   
22                          The\nMerger                 st   
23                                    .                 st   
24                                 2.01                 sn   
25                                 -12-                  n   
26                 Effect of the Merger               ssst   
27                                    .               ssst   
28           

Okay so far looks like we can do logic:
1. Sort by top and then if theres a tiebreaker, sort by left, and then sort by order in the highlight list (in the case of multiple nodes being highlighted in same row with 1 highlight box)
2. If it starts to show that top coordinates are not exact and are approximaate, we are going to need to start doing approximate sorting or some IOU related logic to post process this to infer if something is the same line (which seems extremely likely to work)

In [64]:
def sort_exploded_highlight_box_by_coordinates(exploded_highlight_df):
    '''
    If logic ever has to deal with approximations and some line inference (with IOU math),
    then this function will become more complicated but for now it seems okay
    '''
    return exploded_highlight_df.sort_values(by=['top','left','exploded_highlight_node_order']).reset_index(drop=True)

In [66]:
exploded_highlight_df = sort_exploded_highlight_box_by_coordinates(exploded_highlight_df)
exploded_highlight_df['exploded_highlight_node_order'] = exploded_highlight_df.index.copy()
exploded_highlight_df

highlighted_xpaths  \
0   /html/body/document/type/sequence/filename/des...   
1   /html/body/document/type/sequence/filename/des...   
2   /html/body/document/type/sequence/filename/des...   
3   /html/body/document/type/sequence/filename/des...   
4   /html/body/document/type/sequence/filename/des...   
5   /html/body/document/type/sequence/filename/des...   
6   /html/body/document/type/sequence/filename/des...   
7   /html/body/document/type/sequence/filename/des...   
8   /html/body/document/type/sequence/filename/des...   
9   /html/body/document/type/sequence/filename/des...   
10  /html/body/document/type/sequence/filename/des...   
11  /html/body/document/type/sequence/filename/des...   
12  /html/body/document/type/sequence/filename/des...   
13  /html/body/document/type/sequence/filename/des...   
14  /html/body/document/type/sequence/filename/des...   
15  /html/body/document/type/sequence/filename/des...   
16  /html/body/document/type/sequence/filename/des...   
17  /html/body/document/type/sequence/filename/des...   
18  /html/body/document/type/sequence/filename/des...   
19  /html/body/document/type/sequence/filename/des...   
20  /html/body/document/type/sequence/filename/des...   
21  /html/body/document/type/sequence/filename/des...   
22  /html/body/document/type/sequence/filename/des...   
23  /html/body/document/type/sequence/filename/des...   
24  /html/body/document/type/sequence/filename/des...   
25  /html/body/document/type/sequence/filename/des...   
26  /html/body/document/type/sequence/filename/des...   
27  /html/body/document/type/sequence/filename/des...   
28  /html/body/document/type/sequence/filename/des...   
29  /html/body/document/type/sequence/filename/des...   
30  /html/body/document/type/sequence/filename/des...   
31  /html/body/document/type/sequence/filename/des...   
32  /html/body/document/type/sequence/filename/des...   
33  /html/body/document/type/sequence/filename/des...   
34  /html/body/document/type/sequence/filename/des...   
35  /html/body/document/type/sequence/filename/des...   
36  /html/body/document/type/sequence/filename/des...   
37  /html/body/document/type/sequence/filename/des...   

             highlighted_segmented_text highlighted_labels  \
0          AGREEMENT AND PLAN OF MERGER                  t   
1            dated as of March 22, 2021                  t   
2                        by and between                  t   
3             BANC OF\nCALIFORNIA, INC.                  t   
4                                   and                  t   
5            PACIFIC MERCANTILE BANCORP                  t   
6                     TABLE OF CONTENTS                 st   
7                                   -i-                  n   
8                                  -ii-                  n   
9                              RECITALS                 st   
10                            ARTICLE 1                 sn   
11                  CERTAIN DEFINITIONS                 st   
12                                 1.01                ssn   
13                  Certain Definitions                sst   
14                                    .                sst   
15                                  -2-                  n   
16                                  -3-                  n   
17                                  -4-                  n   
18                                  -5-                  n   
19                                  -6-                  n   
20                            ARTICLE 2                 sn   
21                           THE MERGER                 st   
22                                 2.01                 sn   
23                          The\nMerger                 st   
24                                    .                 st   
25                                 -12-                  n   
26                                  (b)               ssst   
27                   Charter and Bylaws               ssst   
28           

In [67]:
len(df)

1781

In [68]:
merged = pd.merge(df, exploded_highlight_df, left_on='xpaths', right_on='highlighted_xpaths', how='left', indicator=True)
merged

xpaths  \
0                              /html/body/document/type   
1                     /html/body/document/type/sequence   
2            /html/body/document/type/sequence/filename   
3     /html/body/document/type/sequence/filename/des...   
4     /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1787  /html/body/document/type/sequence/filename/des...   
1788  /html/body/document/type/sequence/filename/des...   
1789  /html/body/document/type/sequence/filename/des...   
1790  /html/body/document/type/sequence/filename/des...   
1791  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  \
0                                                EX-2.1                   0   
1                                                     2                   1   
2                                       d35268dex21.htm                   2   
3                                                EX-2.1                   3   
4     EX-2.1 | https://www.sec.gov/Archives/edgar/da...                   4   
...                                                 ...                 ...   
1787                         PACIFIC MERCANTILE BANCORP                1776   
1788                                                By:                1777   
1789                                /s/ Denis Kalscheur                1778   
1790                              Name: Denis Kalscheur                1779   
1791                                    Title: Chairman                1780   

     highlighted_xpaths highlighted_segmented_text highlighted_labels  \
0                   NaN                        NaN                NaN   
1                   NaN                        NaN                NaN   
2                   NaN                        NaN                NaN   
3                   NaN                        NaN                NaN   
4                   NaN                        NaN                NaN   
...                 ...                        ...                ...   
1787                NaN                        NaN                NaN   
1788                NaN                        NaN                NaN   
1789                NaN                        NaN                NaN   
1790                NaN                        NaN                NaN   
1791                NaN                        NaN                NaN   

      segment_number_from_idx highlighted_coordinates  num_entries_1  \
0                         NaN                     NaN            NaN   
1                         NaN                     NaN            NaN   
2                         NaN                     NaN            NaN   
3                         NaN                     NaN            NaN   
4                         NaN                     NaN            NaN   
...                       ...                     ...            ...   
1787                      NaN                     NaN            NaN   
1788                      NaN                     NaN            NaN   
1789                      NaN                     NaN            NaN   
1790                      NaN                     NaN            NaN   
1791                      NaN                     NaN            NaN   

      exploded_highlight_node_order  top  left  width  height     _merge  
0                               NaN  NaN   NaN    NaN     NaN  left_only  
1                               NaN  NaN   NaN    NaN     NaN  left_only  
2                               NaN  NaN   NaN    NaN     NaN  left_only  
3                               NaN  NaN   NaN    NaN     NaN  left_only  
4                               NaN  NaN   NaN    NaN     NaN  left_only  
...                             ...  ...   ...    ...     ...        ...  
1787                            NaN  NaN   NaN    NaN     NaN  left_only  
1788                            NaN  NaN   NaN   

In [69]:
merged['is_outside'] = merged['_merge'].apply(lambda x: 1 if x=='left_only' else 0)

In [70]:
merged.to_csv("drop_check.csv")

In [71]:
#merged[merged['is_outside'] < 1].to_csv('merged_test.csv')

In [72]:
merged[merged['is_outside'] < 1]

xpaths  \
6    /html/body/document/type/sequence/filename/des...   
7    /html/body/document/type/sequence/filename/des...   
8    /html/body/document/type/sequence/filename/des...   
9    /html/body/document/type/sequence/filename/des...   
10   /html/body/document/type/sequence/filename/des...   
11   /html/body/document/type/sequence/filename/des...   
12   /html/body/document/type/sequence/filename/des...   
91   /html/body/document/type/sequence/filename/des...   
190  /html/body/document/type/sequence/filename/des...   
201  /html/body/document/type/sequence/filename/des...   
232  /html/body/document/type/sequence/filename/des...   
233  /html/body/document/type/sequence/filename/des...   
234  /html/body/document/type/sequence/filename/des...   
235  /html/body/document/type/sequence/filename/des...   
236  /html/body/document/type/sequence/filename/des...   
237  /html/body/document/type/sequence/filename/des...   
238  /html/body/document/type/sequence/filename/des...   
271  /html/body/document/type/sequence/filename/des...   
338  /html/body/document/type/sequence/filename/des...   
372  /html/body/document/type/sequence/filename/des...   
416  /html/body/document/type/sequence/filename/des...   
473  /html/body/document/type/sequence/filename/des...   
722  /html/body/document/type/sequence/filename/des...   
723  /html/body/document/type/sequence/filename/des...   
724  /html/body/document/type/sequence/filename/des...   
725  /html/body/document/type/sequence/filename/des...   
726  /html/body/document/type/sequence/filename/des...   
727  /html/body/document/type/sequence/filename/des...   
728  /html/body/document/type/sequence/filename/des...   
733  /html/body/document/type/sequence/filename/des...   
734  /html/body/document/type/sequence/filename/des...   
735  /html/body/document/type/sequence/filename/des...   
736  /html/body/document/type/sequence/filename/des...   
737  /html/body/document/type/sequence/filename/des...   
738  /html/body/document/type/sequence/filename/des...   
739  /html/body/document/type/sequence/filename/des...   
740  /html/body/document/type/sequence/filename/des...   
741  /html/body/document/type/sequence/filename/des...   
742  /html/body/document/type/sequence/filename/des...   
743  /html/body/document/type/sequence/filename/des...   
745  /html/body/document/type/sequence/filename/des...   
746  /html/body/document/type/sequence/filename/des...   
760  /html/body/document/type/sequence/filename/des...   
761  /html/body/document/type/sequence/filename/des...   
762  /html/body/document/type/sequence/filename/des...   
763  /html/body/document/type/sequence/filename/des...   
764  /html/body/document/type/sequence/filename/des...   
765  /html/body/document/type/sequence/filename/des...   
766  /html/body/document/type/sequence/filename/des...   
778  /html/body/document/type/sequence/filename/des...   

                                                  text  all_nodes_ordering  \
6                         AGREEMENT AND PLAN OF MERGER                   6   
7                           dated as of March 22, 2021                   7   
8                                       by and between                   8   
9                            BANC OF\nCALIFORNIA, INC.                   9   
10                                                 and                  10   
11                          PACIFIC MERCANTILE BANCORP                  11   
12                                   TABLE OF CONTENTS                  12   
91                                                 -i-                  91   
190                                               -ii-                 190   
201                                           RECITALS                 201   
232                                          ARTICLE 1                 232   
233                                CERTAIN DEFINITIONS                 233   
234                                               1.01                 234 

In [73]:
def assert_node_ordering_for_merged_table(merged: pd.DataFrame) -> pd.DataFrame:
    '''
    Due to repeat xpaths in both left and right table in merged, need to remove the duplicated rows
    '''
    drop_indices = []
    last_highlight_node_idx = -1
    for i, row in merged[merged['is_outside'] < 1].iterrows():
        if row.exploded_highlight_node_order and row.exploded_highlight_node_order != last_highlight_node_idx + 1:
            drop_indices.append(i)
        else:
            last_highlight_node_idx = row.exploded_highlight_node_order
    merged = merged.drop(drop_indices).reset_index(drop=True)
    
    # Assert both orderings keep their original structure
    assert merged.all_nodes_ordering.is_monotonic_increasing
    assert merged[merged['is_outside'] < 1].exploded_highlight_node_order.is_monotonic_increasing
    
    return merged, drop_indices

In [74]:
merged, drop_indices = assert_node_ordering_for_merged_table(merged)
merged

xpaths  \
0                              /html/body/document/type   
1                     /html/body/document/type/sequence   
2            /html/body/document/type/sequence/filename   
3     /html/body/document/type/sequence/filename/des...   
4     /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1775  /html/body/document/type/sequence/filename/des...   
1776  /html/body/document/type/sequence/filename/des...   
1777  /html/body/document/type/sequence/filename/des...   
1778  /html/body/document/type/sequence/filename/des...   
1779  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  \
0                                                EX-2.1                   0   
1                                                     2                   1   
2                                       d35268dex21.htm                   2   
3                                                EX-2.1                   3   
4     EX-2.1 | https://www.sec.gov/Archives/edgar/da...                   4   
...                                                 ...                 ...   
1775                         PACIFIC MERCANTILE BANCORP                1776   
1776                                                By:                1777   
1777                                /s/ Denis Kalscheur                1778   
1778                              Name: Denis Kalscheur                1779   
1779                                    Title: Chairman                1780   

     highlighted_xpaths highlighted_segmented_text highlighted_labels  \
0                   NaN                        NaN                NaN   
1                   NaN                        NaN                NaN   
2                   NaN                        NaN                NaN   
3                   NaN                        NaN                NaN   
4                   NaN                        NaN                NaN   
...                 ...                        ...                ...   
1775                NaN                        NaN                NaN   
1776                NaN                        NaN                NaN   
1777                NaN                        NaN                NaN   
1778                NaN                        NaN                NaN   
1779                NaN                        NaN                NaN   

      segment_number_from_idx highlighted_coordinates  num_entries_1  \
0                         NaN                     NaN            NaN   
1                         NaN                     NaN            NaN   
2                         NaN                     NaN            NaN   
3                         NaN                     NaN            NaN   
4                         NaN                     NaN            NaN   
...                       ...                     ...            ...   
1775                      NaN                     NaN            NaN   
1776                      NaN                     NaN            NaN   
1777                      NaN                     NaN            NaN   
1778                      NaN                     NaN            NaN   
1779                      NaN                     NaN            NaN   

      exploded_highlight_node_order  top  left  width  height     _merge  \
0                               NaN  NaN   NaN    NaN     NaN  left_only   
1                               NaN  NaN   NaN    NaN     NaN  left_only   
2                               NaN  NaN   NaN    NaN     NaN  left_only   
3                               NaN  NaN   NaN    NaN     NaN  left_only   
4                               NaN  NaN   NaN    NaN     NaN  left_only   
...                             ...  ...   ...    ...     ...        ...   
1775                            NaN  NaN   NaN    NaN     NaN  left_only   
1776                            NaN  NaN 

Why am I losing rows (1 row now) from the left table?

In [75]:
df

xpaths  \
0                              /html/body/document/type   
1                     /html/body/document/type/sequence   
2            /html/body/document/type/sequence/filename   
3     /html/body/document/type/sequence/filename/des...   
4     /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1776  /html/body/document/type/sequence/filename/des...   
1777  /html/body/document/type/sequence/filename/des...   
1778  /html/body/document/type/sequence/filename/des...   
1779  /html/body/document/type/sequence/filename/des...   
1780  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  
0                                                EX-2.1                   0  
1                                                     2                   1  
2                                       d35268dex21.htm                   2  
3                                                EX-2.1                   3  
4     EX-2.1 | https://www.sec.gov/Archives/edgar/da...                   4  
...                                                 ...                 ...  
1776                         PACIFIC MERCANTILE BANCORP                1776  
1777                                                By:                1777  
1778                                /s/ Denis Kalscheur                1778  
1779                              Name: Denis Kalscheur                1779  
1780                                    Title: Chairman                1780  

[1781 rows x 3 columns]

In [76]:
drop_indices

[235, 237, 725, 727, 735, 737, 740, 742, 745, 746, 763, 765]

In [78]:
merged[merged['all_nodes_ordering'] == merged.index]

xpaths  \
0                             /html/body/document/type   
1                    /html/body/document/type/sequence   
2           /html/body/document/type/sequence/filename   
3    /html/body/document/type/sequence/filename/des...   
4    /html/body/document/type/sequence/filename/des...   
..                                                 ...   
732  /html/body/document/type/sequence/filename/des...   
733  /html/body/document/type/sequence/filename/des...   
734  /html/body/document/type/sequence/filename/des...   
735  /html/body/document/type/sequence/filename/des...   
736  /html/body/document/type/sequence/filename/des...   

                                                  text  all_nodes_ordering  \
0                                               EX-2.1                   0   
1                                                    2                   1   
2                                      d35268dex21.htm                   2   
3                                               EX-2.1                   3   
4    EX-2.1 | https://www.sec.gov/Archives/edgar/da...                   4   
..                                                 ...                 ...   
732  . The Charter and Bylaws of Parent as in effec...                 732   
733                                                (c)                 733   
734                               Effect of the Merger                 734   
735  . At the Effective Time, the effect of the Mer...                 735   
736                                              3-114                 736   

                                    highlighted_xpaths  \
0                                                  NaN   
1                                                  NaN   
2                                                  NaN   
3                                                  NaN   
4                                                  NaN   
..                                                 ...   
732  /html/body/document/type/sequence/filename/des...   
733  /html/body/document/type/sequence/filename/des...   
734  /html/body/document/type/sequence/filename/des...   
735  /html/body/document/type/sequence/filename/des...   
736                                                NaN   

    highlighted_segmented_text highlighted_labels  segment_number_from_idx  \
0                          NaN                NaN                      NaN   
1                          NaN                NaN                      NaN   
2                          NaN                NaN                      NaN   
3                          NaN                NaN                      NaN   
4                          NaN                NaN                      NaN   
..                         ...                ...                      ...   
732                          .               ssst                     21.0   
733                        (c)               sssn                     20.0   
734       Effect of the Merger               ssst                     19.0   
735                          .               ssst                     19.0   
736                        NaN                NaN                      NaN   

                     highlighted_coordinates  num_entries_1  \
0                                        NaN            NaN   
1                                        NaN            NaN   
2                                        NaN            NaN   
3                                        NaN            NaN   
4                                        NaN            NaN   
..                                       ...            ...   
732  [8490.12px, 635.828px, 109.219px, 15px]            2.0   
733  [8562.12px, 616.938px, 14.7969px, 15px]            1.0   
734  [8562.12px, 635.078px, 111.312px, 15px]            2.0   
735  [8562.12px, 635.078px, 111.312px, 15px]            2.0   
736                                      NaN            NaN   

     exploded_highlight

In [79]:
merged[merged['all_nodes_ordering'] > merged.index]

xpaths  \
737   /html/body/document/type/sequence/filename/des...   
738   /html/body/document/type/sequence/filename/des...   
739   /html/body/document/type/sequence/filename/des...   
740   /html/body/document/type/sequence/filename/des...   
741   /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1775  /html/body/document/type/sequence/filename/des...   
1776  /html/body/document/type/sequence/filename/des...   
1777  /html/body/document/type/sequence/filename/des...   
1778  /html/body/document/type/sequence/filename/des...   
1779  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  \
737                                                2.02                 738   
738                             Closing; Effective Time                 739   
739                  . The closing of the Merger (the “                 740   
740                                             Closing                 741   
741   ”) shall take place\nremotely by mutual exchan...                 742   
...                                                 ...                 ...   
1775                         PACIFIC MERCANTILE BANCORP                1776   
1776                                                By:                1777   
1777                                /s/ Denis Kalscheur                1778   
1778                              Name: Denis Kalscheur                1779   
1779                                    Title: Chairman                1780   

     highlighted_xpaths highlighted_segmented_text highlighted_labels  \
737                 NaN                        NaN                NaN   
738                 NaN                        NaN                NaN   
739                 NaN                        NaN                NaN   
740                 NaN                        NaN                NaN   
741                 NaN                        NaN                NaN   
...                 ...                        ...                ...   
1775                NaN                        NaN                NaN   
1776                NaN                        NaN                NaN   
1777                NaN                        NaN                NaN   
1778                NaN                        NaN                NaN   
1779                NaN                        NaN                NaN   

      segment_number_from_idx highlighted_coordinates  num_entries_1  \
737                       NaN                     NaN            NaN   
738                       NaN                     NaN            NaN   
739                       NaN                     NaN            NaN   
740                       NaN                     NaN            NaN   
741                       NaN                     NaN            NaN   
...                       ...                     ...            ...   
1775                      NaN                     NaN            NaN   
1776                      NaN                     NaN            NaN   
1777                      NaN                     NaN            NaN   
1778                      NaN                     NaN            NaN   
1779                      NaN                     NaN            NaN   

      exploded_highlight_node_order  top  left  width  height     _merge  \
737                             NaN  NaN   NaN    NaN     NaN  left_only   
738                             NaN  NaN   NaN    NaN     NaN  left_only   
739                             NaN  NaN   NaN    NaN     NaN  left_only   
740                             NaN  NaN   NaN    NaN     NaN  left_only   
741                             NaN  NaN   NaN    NaN     NaN  left_only   
...                             ...  ...   ...    ...     ...        ...   
1775                            NaN  NaN   NaN    NaN     NaN  left_only   
1776                            NaN  NaN 

Okay I see what happened. At index 737 for all nodes ordering, somehow the join had a double repeat. All examples of that node in the left dataframe were deleted since 2 highlight nodes before it matched with it. Not sure why tho?

The (c) node was already matched as well as what followed it but then it and one of the successors was matched again. I think there can be some kind of check to see if highlight node was already matched and if removing current row would destroy the all nodes order, just dont delete the row... 

Seems weird but maybe it would work. 

Maybe another method could be tracking the start coordinates of the text in each node and then somehow getting a stricter order on that...

Or maybe tracking which char offset from node beginning a highlight belongs to. then if theres overlap in the inveral, we merge, if not we move on. do the same for the all nodes as well

Let's do the BIES tagging now

In [14]:
def tag_bies_for_highlights(merged: pd.DataFrame) -> pd.DataFrame:
    
    tags = []
    count = 1
    for i, row in merged.iterrows():
        list_entry_count = row.num_entries_1
        
        # Non highlighted row
        if math.isnan(list_entry_count):
            tags.append('o')
        
        # Single highlighted node
        elif list_entry_count == 1:
            tags.append(f's_{row.highlighted_labels}')
            count = 1
        
        # Last entry in group greater than size 1
        elif count == list_entry_count:
            tags.append(f'e_{row.highlighted_labels}')
            count = 1
        elif (count < list_entry_count) and count == 1:
            tags.append(f'b_{row.highlighted_labels}')
            count += 1
        elif (count < list_entry_count) and count > 1:
            tags.append(f'i_{row.highlighted_labels}')
            count += 1
    merged['tagged_sequence'] = tags
    return merged

In [15]:
merged = tag_bies_for_highlights(merged)
merged

xpaths  \
0                              /html/body/document/type   
1                     /html/body/document/type/sequence   
2            /html/body/document/type/sequence/filename   
3     /html/body/document/type/sequence/filename/des...   
4     /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1751  /html/body/document/type/sequence/filename/des...   
1752  /html/body/document/type/sequence/filename/des...   
1753  /html/body/document/type/sequence/filename/des...   
1754  /html/body/document/type/sequence/filename/des...   
1755  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  \
0                                                EX-2.1                   0   
1                                                     2                   1   
2                                       d35268dex21.htm                   2   
3                                                EX-2.1                   3   
4     EX-2.1 | https://www.sec.gov/Archives/edgar/da...                   4   
...                                                 ...                 ...   
1751                         PACIFIC MERCANTILE BANCORP                1776   
1752                                                By:                1777   
1753                                /s/ Denis Kalscheur                1778   
1754                              Name: Denis Kalscheur                1779   
1755                                    Title: Chairman                1780   

     highlighted_xpaths highlighted_segmented_text highlighted_labels  \
0                   NaN                        NaN                NaN   
1                   NaN                        NaN                NaN   
2                   NaN                        NaN                NaN   
3                   NaN                        NaN                NaN   
4                   NaN                        NaN                NaN   
...                 ...                        ...                ...   
1751                NaN                        NaN                NaN   
1752                NaN                        NaN                NaN   
1753                NaN                        NaN                NaN   
1754                NaN                        NaN                NaN   
1755                NaN                        NaN                NaN   

      segment_number  num_entries_1  exploded_highlight_node_order     _merge  \
0                NaN            NaN                            NaN  left_only   
1                NaN            NaN                            NaN  left_only   
2                NaN            NaN                            NaN  left_only   
3                NaN            NaN                            NaN  left_only   
4                NaN            NaN                            NaN  left_only   
...              ...            ...                            ...        ...   
1751             NaN            NaN                            NaN  left_only   
1752             NaN            NaN                            NaN  left_only   
1753             NaN            NaN                            NaN  left_only   
1754             NaN            NaN                            NaN  left_only   
1755             NaN            NaN                            NaN  left_only   

      is_outside tagged_sequence  
0              1               o  
1              1               o  
2              1               o  
3              1               o  
4              1               o  
...          ...             ...  
1751           1               o  
1752           1               o  
1753           1               o  
1754           1               o  
1755           1               o  

[1756 rows x 12 columns]

In [16]:
merged[merged['is_outside'] < 1]

xpaths  \
12    /html/body/document/type/sequence/filename/des...   
13    /html/body/document/type/sequence/filename/des...   
14    /html/body/document/type/sequence/filename/des...   
15    /html/body/document/type/sequence/filename/des...   
16    /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1745  /html/body/document/type/sequence/filename/des...   
1746  /html/body/document/type/sequence/filename/des...   
1747  /html/body/document/type/sequence/filename/des...   
1748  /html/body/document/type/sequence/filename/des...   
1749  /html/body/document/type/sequence/filename/des...   

                                                   text  all_nodes_ordering  \
12                                    TABLE OF CONTENTS                  12   
13                                                 Page                  13   
14                                             RECITALS                  14   
15                                            ARTICLE 1                  15   
16                                  CERTAIN DEFINITIONS                  16   
...                                                 ...                 ...   
1745  IN WITNESS WHEREOF, the Parties hereto have ca...                1770   
1746                           BANC OF CALIFORNIA, INC.                1771   
1747                                                By:                1772   
1748                                    /s/ Jared Wolff                1773   
1749                                  Name: Jared Wolff                1774   

                                     highlighted_xpaths  \
12    /html/body/document/type/sequence/filename/des...   
13    /html/body/document/type/sequence/filename/des...   
14    /html/body/document/type/sequence/filename/des...   
15    /html/body/document/type/sequence/filename/des...   
16    /html/body/document/type/sequence/filename/des...   
...                                                 ...   
1745  /html/body/document/type/sequence/filename/des...   
1746  /html/body/document/type/sequence/filename/des...   
1747  /html/body/document/type/sequence/filename/des...   
1748  /html/body/document/type/sequence/filename/des...   
1749  /html/body/document/type/sequence/filename/des...   

                             highlighted_segmented_text highlighted_labels  \
12                                    TABLE OF CONTENTS                 st   
13                                                 Page                 st   
14                                             RECITALS                 st   
15                                            ARTICLE 1                 st   
16                                  CERTAIN DEFINITIONS                 st   
...                                                 ...                ...   
1745  IN WITNESS WHEREOF, the Parties hereto have ca...                  n   
1746                           BANC OF CALIFORNIA, INC.                  n   
1747                                                By:                  n   
1748                                    /s/ Jared Wolff                  n   
1749                                      Name: Jared W                  n   

      segment_number  num_entries_1  exploded_highlight_node_order _merge  \
12               0.0           15.0                            0.0   both   
13               0.0           15.0                            1.0   both   
14               0.0           15.0                            2.0   both   
15               0.0           15.0                            3.0   both   
16               0.0           15.0                            4.0   both   
...              ...            ...                            ...    ...   
1745            28.0            5.0                          285.0   both   
1746            28.0            5.0                          286.0   both   
1747            28.0            5.0                 

In [17]:
merged.to_csv('overlabeled_labeled.csv')

I think I need to modify this algorithm. The overlabeled file that I just tried somehow labeled index 245 as b_n but i think its a single highlight node? It looks like the master nodes might be traversing in a different order tbh not sure, I did label pretty wildly... but still not sure how that happens?

In [20]:
highlighted_df.iloc[2]

highlighted_xpaths            [/html/body/document/type/sequence/filename/de...
highlighted_segmented_text    [” has the meaning set forth in Section 6.03(d...
highlighted_labels                                                            n
segment_number                                                                2
num_entries_1                                                                22
num_entries_2                                                                22
Name: 2, dtype: object

In [21]:
highlighted_df.iloc[2]['highlighted_xpaths']

['/html/body/document/type/sequence/filename/description/text/center[5]/div/p[7]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[8]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[8]/i',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[8]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[9]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[9]/i',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[9]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[10]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[10]/i',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[10]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p[11]',
 '/html/body/document/type/sequence/filename/description/text/center[5]/div/p